pkg/admission/prep_downscale.go

```diff
@@ -0,0 +1,246 @@
+	if oldReplicas == nil && newReplicas == nil {
+		level.Debug(logger).Log("msg", "no replicas change, allowing")
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+	// Changes from/to nil scale are not downscales strictly speaking.
+	if oldReplicas == nil || newReplicas == nil {
+		return allowWarn(logger, "old/new replicas is nil, allowing the change")
+	}
+	// If it's not a downscale, just log debug.
+	if *oldReplicas < *newReplicas {
+		level.Debug(logger).Log("msg", "upscale allowed")
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+	if *oldReplicas == *newReplicas {
+		level.Debug(logger).Log("msg", "no replicas change, allowing")
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+
+	// Get the resource labels: for example, for a StatefulSet, it will be the labels of the StatefulSet itself,
+	// while for a Scale object, it will be the labels of the StatefulSet/Deployment/ReplicaSet that the Scale object belongs to.
+	var lbls map[string]string
+	switch o := oldObj.(type) {
+	case *appsv1.Deployment:
+		lbls = o.Labels
+	case *appsv1.StatefulSet:
+		lbls = o.Labels
+	case *appsv1.ReplicaSet:
+		lbls = o.Labels
+	case *autoscalingv1.Scale:
+		lbls, err = getResourceLabels(ctx, ar, api)
+		if err != nil {
+			return allowBecauseCannotGetResource(ar, logger, err)
+		}
+	default:
+		return allowWarn(logger, fmt.Sprintf("unsupported type %T, allowing the change", o))
+	}
+
+	if lbls[PrepDownscaleLabelKey] != PrepDownscaleLabelValue {
+		// Not labeled, nothing to do.
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+
+	port := lbls[PrepDownscalePortKey]
+	if port == "" {
+		reviewResponse := v1.AdmissionResponse{
+			Allowed: false,
+			Result: &metav1.Status{
+				Message: fmt.Sprintf("downscale of %s/%s in %s from %d to %d replicas is not allowed because the %v label is not set or empty.", ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas, PrepDownscalePortKey),
+			},
+		}
+		level.Warn(logger).Log("msg", fmt.Sprintf("downscale not allowed because the %v label is not set or empty", PrepDownscalePortKey))
+		return &reviewResponse
+	}
+
+	path := lbls[PrepDownscalePathKey]
+	if path == "" {
+		reviewResponse := v1.AdmissionResponse{
+			Allowed: false,
+			Result: &metav1.Status{
+				Message: fmt.Sprintf("downscale of %s/%s in %s from %d to %d replicas is not allowed because the %v label is not set or empty.", ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas, PrepDownscalePathKey),
+			},
+		}
+		level.Warn(logger).Log("msg", fmt.Sprintf("downscale not allowed because the %v label is not set or empty", PrepDownscalePathKey))
+		return &reviewResponse
+	}
+
+	if *ar.Request.DryRun {
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+
+	rolloutGroup := lbls[RolloutGroupLabelKey]
+	if rolloutGroup == "" {
+		// Not part of a rollout group, nothing to do.
+		return &v1.AdmissionResponse{Allowed: true}
+	}
```

Comment:
I think this is not correct. Someone labeled this statefulset as it should be prepared for downscale. It's not part of a rollout group, so there's no need to check other zones, but we still want to prepare it for downscale.
I'd say that:
```
if rolloutGroup != "" {
    // ... check other zones last downscale
}
```

Reply:
Indeed, that is a scenario I hadn't thought off.